tools/src/main/scala/scala/scalanative/build/NativeSourcesCompilerPlugin.scala

```diff
@@ -0,0 +1,282 @@
+package scala.scalanative
+package build
+
+import java.io.File
+import java.nio.file.{Files, Path, Paths}
+import scalanative.build.IO.RichPath
+import scalanative.build.Mode._
+
+import NativeSourcesCompilerPlugin._
+
+/** A plugin interface for compiling native sources, such as C, C++, and Rust.
+ */
+trait NativeSourcesCompilerPlugin {
+
+  /** The name of this plugin */
+  def name: String
+
+  /** The extensions this plugin accepts. e.g. `.c` */
+  def extensions: Seq[String]
+
+  /** Return a command to compile the given file and the result.
+   */
+  def compile(config: Config, path: Path): Option[CompilationContext]
+
+}
+
+object NativeSourcesCompilerPlugin {
+
+  /** The command to compile a file and the result.
+   */
+  case class CompilationContext(
+      command: Seq[String],
+      result: CompilationResult
+  )
+
+}
+
+sealed abstract class ClangSourcesCompilerPlugin
+    extends NativeSourcesCompilerPlugin {
+
+  protected def stdflag(config: Config): Seq[String]
+
+  protected def compiler(config: Config): String = config.clang.abs
+
+  protected def forceRebuild(input: Path): Boolean = false
+
+  private def opt(config: Config): String =
+    config.mode match {
+      case Mode.Debug       => "-O0"
+      case Mode.ReleaseFast => "-O2"
+      case Mode.ReleaseFull => "-O3"
+    }
+
+  def compile(config: Config, inpath: Path): Option[CompilationContext] = {
```

Comment:
Could you explain why this returns an option because in the C/C++ plugin it seems to be not needed?

Reply:
That's a good question. I'm actually not 100% sure, this was the existing method signature. The rust compiler does return `None` in some instances. It seems to indicate that there is nothing to be done. But I'm not sure what differentiates it from returning a `CompilationContext` with an empty command. @WojciechMazur will probably know :)

https://github.com/scala-native/scala-native/blob/653dfabe2ebbfde9495e861f4166ae3e98be8c9d/tools/src/main/scala/scala/scalanative/build/NativeSourcesCompilerPlugin.scala#L183